File: modules/076-skills-management-system/tools/scripts/cmd/list.go
package cmd

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"

	"github.com/vibecoding/skills-cli/internal/config"
	"github.com/vibecoding/skills-cli/internal/gitops"
	"github.com/vibecoding/skills-cli/internal/manifest"
)

type skillJSON struct {
	Name        string `json:"name"`
	Active      bool   `json:"active"`
	Description string `json:"description,omitempty"`
	Owner       string `json:"owner,omitempty"`
}

// RunList handles the `skills list` command.
func RunList(args []string) {
	fs := flag.NewFlagSet("list", flag.ExitOnError)
	verbose := fs.Bool("verbose", false, "Show description and owner from info.json")
	jsonOut := fs.Bool("json", false, "Output skills as JSON array")
	fs.Usage = func() {
		fmt.Print(`List all available skills in the repository.

Usage:
  skills list [--verbose] [--json]

Flags:
`)
		fs.PrintDefaults()
		fmt.Print(`
Active skills (checked out in this workspace) are marked with ✅.
Other skills exist in the repo but are not part of your current groups.

`)
	}

	if err := fs.Parse(args); err != nil {
		os.Exit(1)
	}

	cfg, err := config.Load()
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
		os.Exit(1)
	}

	allSkills, err := gitops.ListAllSkills(cfg.RepoPath())
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: failed to list skills: %v\n", err)
		os.Exit(1)
	}

	// Resolve active skills dynamically from manifests
	groups := ResolveEffectiveGroups(cfg)
	resolvedSkills, _ := manifest.ResolveSkills(cfg.RepoPath(), groups)
	resolvedSkills = ApplyExtraAndExcluded(resolvedSkills, cfg)
	activeSet := make(map[string]bool, len(resolvedSkills))
	for _, s := range resolvedSkills {
		activeSet[s] = true
	}

	// JSON output mode
	if *jsonOut {
		var items []skillJSON
		for _, s := range allSkills {
			item := skillJSON{Name: s, Active: activeSet[s]}
			if info := gitops.LoadSkillInfo(cfg.RepoPath(), s); info != nil {
				item.Description = info.Description
				item.Owner = info.Owner
			}
			items = append(items, item)
		}
		data, _ := json.MarshalIndent(items, "", "  ")
		fmt.Println(string(data))
		return
	}

	// Normal / verbose text output
	fmt.Printf("Skills repository: %s\n", cfg.RepoURL)
	fmt.Printf("Groups:           %v\n\n", cfg.Groups)

	activeCount := 0
	for _, s := range allSkills {
		if activeSet[s] {
			fmt.Printf("  ✅ %s\n", s)
			activeCount++
		} else {
			fmt.Printf("  ○  %s\n", s)
		}
		if *verbose {
			if info := gitops.LoadSkillInfo(cfg.RepoPath(), s); info != nil {
				fmt.Printf("     %s\n", info.Description)
				fmt.Printf("     Owner: %s\n", info.Owner)
			}
		}
	}

	fmt.Printf("\nActive: %d  |  Total: %d\n", activeCount, len(allSkills))
}

File: modules/076-skills-management-system/tools2/src/commands/list.ts
import * as config from '../lib/config';
import * as gitops from '../lib/gitops';
import * as manifest from '../lib/manifest';
import { resolveEffectiveGroups, applyExtraAndExcluded } from './toggle';

export function runList(args: string[]): void {
  if (args.includes('--help') || args.includes('-h')) {
    console.log(`List all available skills in the repository.

Usage:
  skills list [--verbose] [--json]

Flags:
  --verbose  Show description and owner from info.json
  --json     Output skills as JSON array

Active skills (checked out in this workspace) are marked with ✅.
Other skills exist in the repo but are not part of your current groups.

`);
    return;
  }

  const verbose = args.includes('--verbose');
  const jsonOut = args.includes('--json');

  let cfg: config.Config;
  try {
    cfg = config.load();
  } catch (err) {
    console.error(String(err));
    process.exit(1);
  }
  const repoDir = config.REPO_SUB_DIR;

  let allSkills: string[];
  try {
    allSkills = gitops.listAllSkills(repoDir);
  } catch (err) {
    console.error(`Error: failed to list skills: ${err}`);
    process.exit(1);
  }

  // Resolve active skills dynamically from manifests
  const groups = resolveEffectiveGroups(cfg);
  let resolvedSkills: string[];
  try {
    resolvedSkills = manifest.resolveSkills(repoDir, groups);
  } catch {
    resolvedSkills = [];
  }
  resolvedSkills = applyExtraAndExcluded(resolvedSkills, cfg);
  const activeSet = new Set<string>(resolvedSkills);

  // JSON output mode
  if (jsonOut) {
    const items = allSkills.map(s => {
      const entry: Record<string, unknown> = { name: s, active: activeSet.has(s) };
      const info = gitops.loadSkillInfo(repoDir, s);
      if (info) {
        entry.description = info.description;
        entry.owner = info.owner;
      }
      return entry;
    });
    console.log(JSON.stringify(items, null, 2));
    return;
  }

  // Normal / verbose text output
  console.log(`Skills repository: ${cfg.repo_url}`);
  console.log(`Groups:           ${cfg.groups.join(', ')}\n`);

  let activeCount = 0;
  for (const s of allSkills) {
    if (activeSet.has(s)) {
      console.log(`  ✅ ${s}`);
      activeCount++;
    } else {
      console.log(`  ○  ${s}`);
    }
    if (verbose) {
      const info = gitops.loadSkillInfo(repoDir, s);
      if (info) {
        console.log(`     ${info.description}`);
        console.log(`     Owner: ${info.owner}`);
      }
    }
  }

  console.log(`\nActive: ${activeCount}  |  Total: ${allSkills.length}`);
}
